tests/client.rs

```diff
@@ -2246,6 +2246,130 @@ mod conn {
+            let listener = TkTcpListener::bind(SocketAddr::from(([127, 0, 0, 1], 0)))
+                .await
+                .unwrap();
+            let addr = listener.local_addr().unwrap();
+
+            let server = async move {
+                let mut sock = listener.accept().await.unwrap().0;
+                let mut buf = [0; 4096];
+                let n = sock.read(&mut buf).await.expect("read 1");
+
+                // Notably:
+                // - Just a path, since just a path was set
+                // - No host, since no host was set
+                let expected = "GET /a HTTP/1.1\r\n\r\n";
+                assert_eq!(s(&buf[..n]), expected);
+
+                sock.write_all(b"HTTP/1.1 200 OK\r\nContent-Length: 0\r\n\r\n")
+                    .await
+                    .unwrap();
+            };
+
+            let client = async move {
+                let tcp = tcp_connect(&addr).await.expect("connect");
+                let (mut client, conn) = conn::http1::handshake(tcp).await.expect("handshake");
+
+                tokio::task::spawn(async move {
+                    conn.await.expect("http conn");
+                });
+
+                let req: Request<Body> = Request::builder()
+                    .uri("/a")
+                    .body(Default::default())
+                    .unwrap();
+                let mut res = client.send_request(req).await.expect("send_request");
+                assert_eq!(res.status(), hyper::StatusCode::OK);
+                assert!(res.body_mut().next().await.is_none());
+            };
+
+            future::join(server, client).await;
+        }
+
+        #[tokio::test]
+        async fn http2_detect_conn_eof() {
+            use futures_util::future;
+            use hyper::service::{make_service_fn, service_fn};
+            use hyper::{Response, Server};
+
+            let _ = pretty_env_logger::try_init();
+
+            let server = Server::bind(&([127, 0, 0, 1], 0).into())
+                .http2_only(true)
+                .serve(make_service_fn(|_| async move {
+                    Ok::<_, hyper::Error>(service_fn(|_req| {
+                        future::ok::<_, hyper::Error>(Response::new(Body::empty()))
+                    }))
+                }));
+            let addr = server.local_addr();
+            let (shdn_tx, shdn_rx) = oneshot::channel();
+            tokio::task::spawn(async move {
+                server
+                    .with_graceful_shutdown(async move {
+                        let _ = shdn_rx.await;
+                    })
+                    .await
+                    .expect("server")
+            });
+
+            struct TokioExec;
+            impl<F> hyper::rt::Executor<F> for TokioExec
+            where
+                F: std::future::Future + Send + 'static,
+                F::Output: Send + 'static,
+            {
+                fn execute(&self, fut: F) {
+                    tokio::spawn(fut);
+                }
+            }
```

Comment:
Nit: should we move this outside the test (but inside the backports test module)?